File: LintCode_Python/sortList.py
# Sort a linked list in O(n log n) time using constant space complexity.

# Have you met this question in a real interview? Yes
# Example
# Given 1-3->2->null, sort it to 1->2->3->null.


from ListNode import ListNode

# Method2: quick sort
def sortListI(head):
    if head == None or head.next == None:
        return head
    mid = findMiddle(head)
    head = partition(head, mid.val)
    right = sortListI(mid.next)
    mid.next = None
    left = sortListI(head)
    mid.next = right
    return left

# partition list to <, = and >
def partition(head, x):
    if head == None or head.next == None:
        return head
    dummyLeft = ListNode(0)
    lastLeftNode = dummyLeft
    dummyMid = ListNode(0)
    lastMidNode = dummyMid
    dummyRight = ListNode(0)
    lastRightNode = dummyRight
    while head != None:
        if head.val < x:
            lastLeftNode.next = head
            lastLeftNode = lastLeftNode.next
        elif head.val == x:
            lastMidNode.next = head
            lastMidNode = lastMidNode.next
        else:
            lastRightNode.next = head
            lastRightNode = lastRightNode.next
        head = head.next
    lastLeftNode.next = dummyMid.next
    lastMidNode.next = dummyRight.next
    lastRightNode.next = None
    return dummyLeft.next



# Method1: merge sort
def sortList(head):
    if head == None or head.next == None:
        return head
    mid = findMiddle(head)
    right = sortList(mid.next)
    mid.next = None
    left = sortList(head)
    return merge(left, right)


# helper: find the middle of linked list
def findMiddle(head):
    if head == None or head.next == None:
        return head
    slow = head
    fast = head.next
    while fast != None and fast.next != None:
        slow = slow.next
        fast = fast.next.next
    return slow


# helper: merge two sorted linked list
def merge(head1, head2):
    if head1 == None and head2 == None:
        return None

    dummy = ListNode(0)
    lastNode = dummy
    while head1 != None and head2 != None:
        if head1.val < head2.val:
            lastNode.next = head1
            head1 = head1.next
        else:
            lastNode.next = head2
            head2 = head2.next
        lastNode = lastNode.next
    if head1 != None:
        lastNode.next = head1
    else:
        lastNode.next = head2
    return dummy.next


head = ListNode.arrayToList([1, 3, 2, 4, 5, 7, 6])
ListNode.printList(sortList(head))
ListNode.printList(sortListI(head))

File: LeetCode_Java/src/SortList.java
/**
 * Sort a linked list in O(n log n) time using constant space complexity.
 * 
 * Tags: LinkedList, Sort
 * @author wendi
 *
 */
public class SortList {

	/**
	 * MergeSort
	 * Using helper function to find the mid node of list, then merging sort head list and mid list.
	 * @param ListNode head
	 * @return ListNode
	 * Time: O(n)
	 * Space: O(1)
	 * Stack space: O(log(n))
	 */
	public ListNode sortList(ListNode head) {
		if (head == null || head.next == null) return head;
		return helper(head);
	}
	
	private ListNode helper(ListNode head) {
		if (head.next == null) return head;
		ListNode slow = head;
		ListNode fast = head;
		ListNode mid;
		while (fast.next != null && fast.next.next != null) {
			slow = slow.next;
			fast = fast.next.next;
		}
		mid = slow.next;
		slow.next = null;
		return mergeSort(helper(head), helper(mid));
	}
	
	private ListNode mergeSort(ListNode head1, ListNode head2) {
		if (head1 == null) return head2;
		if (head2 == null) return head1;
		ListNode dummy = new ListNode(0);
		ListNode node = dummy;
		while (head1 != null && head2 != null) {
			if (head1.val < head2.val) {
				node.next = head1;
				head1 = head1.next;
			}
			else {
				node.next = head2;
				head2 = head2.next;
			}
			node = node.next;
		}
		node.next = head1 != null ? head1 : head2;
		return dummy.next;
	}
	
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		SortList result = new SortList();
		ListNode head = ListNode.generateLinkedList(new int[] {6, 2, 5, 3, 1, 4, 7});
		ListNode.printLinkedList(head);
		ListNode.printLinkedList(result.sortList(head));
	}

}
